README
Test-BDD-Definitions

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Test::BDD::Definitions

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-BDD-Definitions

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Test-BDD-Definitions

    CPAN Ratings
        http://cpanratings.perl.org/d/Test-BDD-Definitions

    Search CPAN
        http://search.cpan.org/dist/Test-BDD-Definitions/


LICENSE AND COPYRIGHT

Copyright (C) 2018 Mikhail Ivanov

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Test-BDD-Cucumber-Definitions

{{$NEXT}}

0.41 2018-07-05T16:03:43Z

    [Var]
    - In the "Var" module, a step was added to create the variable with the data structure element

0.40 2018-04-23T20:31:20Z

    [ All ]
    - The "Types" module has been removed; All types are moved to the "Definitions" module

0.39 2018-04-22T20:23:51Z

    [ All ]
    - Functions for parameter validation are transferred to the Definitions module

0.38 2018-04-21T20:58:16Z

    [ All ]
    - Perltidy tests was added
    - Perlcritic tests was added

0.37 2018-04-19T21:10:31Z

    [ All ]
    - Сompilation tests were added

0.36 2018-04-19T19:32:36Z

    [Base]
    - Module "Base" was added

    [Struct]
    - In the module "Struct" a step is added to read the result of the query to the database

0.35 2018-04-12T20:08:03Z

    [ Var ]
    - Module "Var" is rewritten in object-oriented style
    - The Var::scenario function was added

0.34 2018-04-10T20:09:17Z

    [ Zip ]
    - Module "Zip" is rewritten in object-oriented style
    - The Zip::member_names function was added

0.33 2018-04-10T19:29:27Z

    [ HTTP ]
    - Module "HTTP" is rewritten in object-oriented style
    - The HTTP::content function was added

0.32 2018-04-09T18:00:19Z

    [ File ]
    - Module "File" is rewritten in object-oriented style
    - The File::content function was added

    [ Struct ]
    - Module "Struct" is rewritten in object-oriented style
    - The expression "array" is replaced with "list"

0.31 2018-04-03T20:24:27Z

    [ Struct ]
    - The Struct::struct_data_element function was added

0.30 2018-04-01T19:01:56Z

    [ File ]
    -  All the data in the File module is stored in the object

    [ Struct ]
    - The step "file content read JSON" is added to the Struct module
    - The Struct module adds steps to check the hash keys

    [ All ]
    - Requirement Moose was added
    - Requirement namespace::autoclean was added

0.29 2018-03-25T18:16:33Z

    [ All ]
    - Error message when checking type was fixed

0.28 2018-03-25T18:07:36Z

    [ All ]
    - Requirement File::Slurper was added
    - Requirement IO::Capture was added

    [ File ]
    - Steps to read the files were added

0.27 2018-03-20T20:29:03Z

    [ All ]
    - The "validator_n" validation function has been added
    - Requirement IPC::Run3 was added

    [ File ]
    - Module File was added

0.26 2018-03-11T20:32:12Z

    [ HTTP ]
    - The body of the HTTP request is encoded in utf8

0.25 2018-03-10T22:07:18Z

    [ All ]
    - The Validator module with a set of basic functions for parameter validation is added
    - All step definitions are placed in the import function

    [ TBCD ]
    - The TBCD module has been added. The module loads all the step definitions at once.

    [ HTTP ]
    - Functions renamed more accurately

0.24 2018-03-06T20:35:27Z

    [ All ]
    - The POD of the main module has been edited

0.23 2018-03-05T21:04:21Z
    [ All ]
    - Migrate to Minilla

0.22    2018-03-04
        [ All ]
        - In all types of Tbcd*, the scenario and environment variables are interpolated

        [ HTTP ]
        - Types have been removed, HTTP module are redesigned to use the ::Types

        [ Struct ]
        - Types have been removed, Struct module are redesigned to use the ::Types
        - In the module Struct were added functions for working with arrays
        - All functions are renamed more uniformly

        [ Zip ]
        - Functions in the Zip module have been renamed

0.21    2018-03-01
        [ All ]
        - Types common to all modules have been added (::Types)
        
        [ HTTP ]
        - The HTTP module uses common types from ::Types
        
        [ Var ]
        - Module Var was added

0.20    2018-02-28
        [ HTTP ]
        - In the values specified in the HTTP module, you can use scenario variables
            (but you can not specify the variables in the step, only manually)

0.19    2018-02-25
        [ Zip ]
        - Steps to work with Zip archives are added
        - Requirements Archive::Zip, IO::String has been added

        [ Struct ]
        - The function for reading data in the JSON format has been moved from the JSON module to the Struct module

        [ JSON ]
        - The JSON module has been removed

0.18    2018-02-24
        [ All ]
        - Requirement MooseX::Types::Common was added
        - Module TypeConstraints is removed. The HTTP and Struct modules have their own types.

        [ HTTP ]
        - Steps was edited

        [ Struct ]
        - Steps was edited

0.17    2018-02-23
        [ HTTP ]
        - URL of the http request can use environment variables

        [ JSON ]
        - POD fixed

0.16    2018-02-21
        [ HTTP ]
        - Accidentally forgotten import function C was added

0.15    2018-02-19
        [ Definitions ]
        - Functions S, C, Given, When and Then are exported from the module Definitions

        [ Struct ]
        - The Struct module stores the data in S->{struct}->{data}

0.14    2018-02-18
        [ JSON ]
        - Module JSON is divided into two separate modules JSON and Struct

        [ Struct ]
        - Module was added

0.13    2018-02-18
        [ HTTP ]
        - The steps in the HTTP::In module have been renamed
        - For the HTTP::Ru module, documentation is written in POD format

        [ JSON ]
        - Module "Data" was renamed to "JSON"
        - For the JSON::Ru module, documentation is written in POD format

0.12    2018-02-11
        [ HTTP ]
        - Module HTTP::Ru with the definitions of steps in the Russian language is added
        - *_decoded functions in module HTTP have been removed

        [ Data ]
        - Some steps in module Data::In renamed

0.11    2018-02-07
        [Data]
        - Module Data::Ru with the definitions of steps in the Russian language is added

0.10    2018-02-07

lib/Test/BDD/Definitions.pm
package Test::BDD::Definitions;

use 5.006;
use strict;
use warnings;

=head1 NAME

Test::BDD::Definitions - The great new Test::BDD::Definitions!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Test::BDD::Definitions;

    my $foo = Test::BDD::Definitions->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Mikhail Ivanov, C<< <m.ivanych at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-test-bdd-definitions at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Test-BDD-Definitions>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Test::BDD::Definitions


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-BDD-Definitions>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Test-BDD-Definitions>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Test-BDD-Definitions>

=item * Search CPAN

L<http://search.cpan.org/dist/Test-BDD-Definitions/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2018 Mikhail Ivanov.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Test::BDD::Definitions


lib/Test/BDD/Cucumber/Definitions.pm
package Test::BDD::Cucumber::Definitions;

use 5.006;
use strict;
use warnings;

use DDP ( show_unicode => 1 );
use Exporter qw(import);
use Moose::Util::TypeConstraints;
use Params::ValidationCompiler qw(validation_for);
use Test::BDD::Cucumber::StepFile qw(Given When Then);
use Test::More;
use Try::Tiny;

=head1 NAME

Test::BDD::Cucumber::Definitions - a collection of step definitions for Test
Driven Development

=head1 VERSION

Version 0.39

=cut

our $VERSION = '0.39';

=head1 SYNOPSIS

In file B<features/step_definitions/tbcd_steps.pl>:

    #!/usr/bin/perl

    use strict;
    use warnings;

    use Test::BDD::Cucumber::Definitions::TBCD::In;

In file B<features/site.feature>:

    Feature: Site
        Site tests

    Scenario: Loading the page
        When http request "GET" send "http://metacpan.org"
        Then http response code eq "200"

... and, finally, in the terminal:

    $ pherkin

      Site
        Site tests

        Scenario: Loading the page
          When http request "GET" send "http://metacpan.org"
          Then http response code eq "200"


=head1 EXPORT

The module exports functions C<S>, C<C>, C<Given>, C<When> and C<Then>.
These functions are identical to the same functions from the module
L<Test::BDD::Cucumber>.

Additionally, the module exports several functions for parameter validation.
These functions are exported by the C<:validator> tag.

By default, no functions are exported. All functions must be imported
explicitly.

=cut

our @EXPORT_OK = qw(
    S C Given When Then
    validator_i
    validator_n
    validator_s
    validator_r
    validator_ni
    validator_ns
    validator_nn
    validator_nr
);

our %EXPORT_TAGS = (
    validator => [
        qw(
            validator_i
            validator_n
            validator_s
            validator_r
            validator_ni
            validator_ns
            validator_nn
            validator_nr
            )
    ]
);

sub S { return Test::BDD::Cucumber::StepFile::S }
sub C { return Test::BDD::Cucumber::StepFile::C }

# Interpolation of variables (scenario and environment)
sub _interpolate {
    my ($value) = @_;

    my $orig = $value;

    # Scenario variables
    my $is = $value =~ s| S\{ (.+?) \} |
        S->{Var}->scenario($1) // q{};
    |gxe;

    # Environment variables
    my $ie = $value =~ s/ \$\{ (.+?) \} /
        $ENV{$1} || '';
    /gxe;

    if ( $is || $ie ) {
        diag( sprintf( q{Inteprolated value "%s" = %s}, $orig, np $value) );
    }

    return $value;
}

# TbcdInt
subtype(
    'TbcdInt',
    as 'Int',
    message {
        sprintf( '%s is not a valid TBCD Int', np $_);
    }
);

coerce(
    'TbcdInt',
    from 'Str',
    via { _interpolate $_ }
);

# TbcdStr
subtype(
    'TbcdStr',
    as 'Str',
    message {
        sprintf( '%s is not a valid TBCD Str', np $_);
    }
);

coerce(
    'TbcdStr',
    from 'Str',
    via { _interpolate $_}
);

# TbcdNonEmptyStr
subtype(
    'TbcdNonEmptyStr',
    as 'Str',
    where { length($_) > 0 },
    message {
        sprintf( '%s is not a valid TBCD NonEmptyStr', np $_);
    }
);

coerce(
    'TbcdNonEmptyStr',
    from 'Str',
    via { _interpolate $_}
);

# TbcdRegexpRef
subtype(
    'TbcdRegexpRef',
    as 'RegexpRef',
    message {
        sprintf( '%s is not a valid TBCD RegexpRef', np $_);
    }
);

coerce(
    'TbcdRegexpRef',
    from 'Str',
    via {
        my $value = _interpolate $_;

        try {
            qr/$value/;    ## no critic [RegularExpressions::RequireExtendedFormatting]
        }
        catch {
            return $value;
        };
    }
);

my $validator_i = validation_for(
    params => [

        # value integer
        { type => find_type_constraint('TbcdInt') },
    ]
);

sub validator_i {
    return $validator_i;
}

my $validator_n = validation_for(
    params => [

        # name
        { type => find_type_constraint('TbcdNonEmptyStr') },
    ]
);

sub validator_n {
    return $validator_n;
}

my $validator_s = validation_for(
    params => [

        # value string
        { type => find_type_constraint('TbcdStr') },
    ]
);

sub validator_s {
    return $validator_s;
}

my $validator_r = validation_for(
    params => [

        # value regexp
        { type => find_type_constraint('TbcdRegexpRef') }
    ]
);

sub validator_r {
    return $validator_r;
}

my $validator_ni = validation_for(
    params => [

        # name
        { type => find_type_constraint('TbcdNonEmptyStr') },

        # value int
        { type => find_type_constraint('TbcdInt') },
    ]
);

sub validator_ni {
    return $validator_ni;
}

my $validator_ns = validation_for(
    params => [

        # name
        { type => find_type_constraint('TbcdNonEmptyStr') },

        # value string
        { type => find_type_constraint('TbcdStr') }
    ]
);

sub validator_ns {
    return $validator_ns;
}

my $validator_nn = validation_for(
    params => [

        # value non empty string
        { type => find_type_constraint('TbcdNonEmptyStr') },

        # value non empty string
        { type => find_type_constraint('TbcdNonEmptyStr') },
    ]
);

sub validator_nn {
    return $validator_nn;
}

my $validator_nr = validation_for(
    params => [

        # name
        { type => find_type_constraint('TbcdNonEmptyStr') },

        # value regexp
        { type => find_type_constraint('TbcdRegexpRef') }
    ]
);

sub validator_nr {
    return $validator_nr;
}

=head1 AUTHOR

Mikhail Ivanov C<< <m.ivanych@gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Mikhail Ivanov.

This is free software; you can redistribute it and/or modify it
under the same terms as the Perl 5 programming language system itself.

=cut

1;


lib/Test/BDD/Cucumber/Definitions/Var.pm
package Test::BDD::Cucumber::Definitions::Var;

use strict;
use warnings;

use DDP ( show_unicode => 1 );
use Exporter qw(import);
use Test::BDD::Cucumber::Definitions qw(S :validator);
use Test::More;

our $VERSION = '0.40';

our @EXPORT_OK = qw(Var);

## no critic [Subroutines::RequireArgUnpacking]

sub Var {
    return __PACKAGE__;
}

sub scenario_var_set {
    my $self = shift;
    my ( $name, $value ) = validator_ns->(@_);

    S->{Var} = __PACKAGE__;

    S->{_Var}->{scenario}->{vars}->{$name} = $value;

    return 1;
}

sub scenario_var_random {
    my $self = shift;
    my ( $name, $length ) = validator_ni->(@_);

    S->{Var} = __PACKAGE__;

    my @CHARS = ( 'a' .. 'z', 'A' .. 'Z', 0 .. 9 );
    my $str = "X" x $length;
    $str =~ s/X(?=X*\z)/$CHARS[ int( rand( @CHARS ) ) ]/gex;

    S->{_Var}->{scenario}->{vars}->{$name} = $str;

    return 1;
}

sub scenario_var_struct {
    my $self = shift;
    my ( $name, $jsonpath ) = validator_ns->(@_);

    S->{Var} = __PACKAGE__;

    my $value = S->{Struct}->data_element($jsonpath);

    S->{_Var}->{scenario}->{vars}->{$name} = $value;

    return 1;
}

sub scenario {
    my $self = shift;
    my ($name) = validator_n->(@_);

    return S->{_Var}->{scenario}->{vars}->{$name};
}

1;


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Test::BDD::Cucumber::Definitions',
    AUTHOR           => q{Mikhail Ivanov <m.ivanych@gmail.com>},
    VERSION_FROM     => 'lib/Test/BDD/Cucumber/Definitions.pm',
    ABSTRACT_FROM    => 'lib/Test/BDD/Cucumber/Definitions.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'Const::Fast'                   => 0,
        'DDP'                           => 0,
        'Hash::MultiValue'              => 0,
        'HTTP::Response'                => 0,
        'HTTP::Tiny'                    => 0,
        'JSON::Path'                    => 0,
        'JSON::XS'                      => 0,
        'MooseX::Types'                 => 0,
        'MooseX::Types::Common'         => 0,
        'Params::ValidationCompiler'    => '0.22',
        'Test::BDD::Cucumber'           => 0,
        'Try::Tiny'                     => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            license     => [ 'http://dev.perl.org/licenses/' ],
            bugtracker  => {
                web    => 'https://github.com/ivanych/Test-BDD-Cucumber-Definitions/issues',
            },
            repository => {
                type => 'git',
                url  => 'https://github.com/ivanych/Test-BDD-Cucumber-Definitions.git',
                web  => 'https://github.com/ivanych/Test-BDD-Cucumber-Definitions',
            },
        },
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Test-BDD-Cucumber-Definitions-*' },
);
